ledger/execution/tx_coinbase.go

```diff
@@ -116,16 +123,77 @@ func (exec *CoinbaseTxExecutor) process(chainID string, view *st.StoreView, tran
 }
 
 // CalculateReward calculates the block reward for each account
-func CalculateReward(view *st.StoreView, validatorAddresses []common.Address) map[string]types.Coins {
+func CalculateReward(view *st.StoreView, validatorSet *core.ValidatorSet) map[string]types.Coins {
 	accountReward := map[string]types.Coins{}
+	blockHeight := view.Height() + 1 // view points to the parent block
+	if blockHeight < common.HeightEnableValidatorReward {
+		grantValidatorsWithZeroReward(validatorSet, &accountReward)
+	} else {
+		grantStakerReward(view, validatorSet, &accountReward, blockHeight)
+	} // TODO: calculate reward for the guardian nodes' stakers
+
+	return accountReward
+}
 
-	for _, validatorAddress := range validatorAddresses {
-		// Initial Mainnet release should not reward the validators until the guardians ready to deploy
-		zeroReward := types.Coins{}.NoNil()
-		accountReward[string(validatorAddress[:])] = zeroReward
+func grantValidatorsWithZeroReward(validatorSet *core.ValidatorSet, accountReward *map[string]types.Coins) {
+	// Initial Mainnet release should not reward the validators until the guardians ready to deploy
+	zeroReward := types.Coins{}.NoNil()
+	for _, v := range validatorSet.Validators() {
+		(*accountReward)[string(v.Address[:])] = zeroReward
 	}
+}
 
-	return accountReward
+func grantStakerReward(view *st.StoreView, validatorSet *core.ValidatorSet, accountReward *map[string]types.Coins, blockHeight uint64) {
+	isACheckpoint := blockHeight%uint64(checkpointInterval) == 0
+	if !isACheckpoint {
+		return
+	}
+
+	totalStake := validatorSet.TotalStake()
+	if totalStake.Cmp(big.NewInt(0)) != 0 {
+
+		stakeSourceMap := map[common.Address]*big.Int{}
+
+		// TODO - Need to confirm: should we get the VCP from the current view? What if there is a stake deposit/withdraw?
+		vcp := view.GetValidatorCandidatePool()
+		for _, v := range validatorSet.Validators() {
+			validatorAddr := v.Address
+			stakeDelegate := vcp.FindStakeDelegate(validatorAddr)
+			if stakeDelegate == nil { // should not happen
+				panic(fmt.Sprintf("Failed to find stake delegate in the VCP: %v", hex.EncodeToString(validatorAddr[:])))
+			}
+
+			stakes := stakeDelegate.Stakes
+			for _, stake := range stakes {
+				if stake.Withdrawn {
+					continue
+				}
+				stakeAmount := stake.Amount
+				stakeSource := stake.Source
+				if stakeAmountSum, exists := stakeSourceMap[stakeSource]; exists {
+					stakeAmountSum := big.NewInt(0).Add(stakeAmountSum, stakeAmount)
+					stakeSourceMap[stakeSource] = stakeAmountSum
+				} else {
+					stakeSourceMap[stakeSource] = stakeAmount
+				}
+			}
+		}
+
+		// the source of the stake divides the block reward proportional to their stake
+		totalReward := big.NewInt(1).Mul(tfuelRewardPerBlock, big.NewInt(checkpointInterval))
+		for stakeSourceAddr, stakeAmountSum := range stakeSourceMap {
```

Comment:
Does the iteration order matter? Note: Golang does not guarantee the map iteration order

Same for ledger.go:551 (in adCoinbaseTx)